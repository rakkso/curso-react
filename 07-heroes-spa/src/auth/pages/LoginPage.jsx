import React, { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {
  const navigate = useNavigate();

  const { login } = useContext (AuthContext);

  const lastPath  = localStorage.getItem('lastPath') || '/';


  const onLogin =() =>{
    
    navigate(lastPath,{replace:true});
    console.log ("LoginPage haciendo login");
    login("Ã“scar Bella");
  }

  return (
    <div className='container mt-5'>
        <h1>LoginPage</h1>
        <button onClick={onLogin} className='btn btn-primary'>Login</button>
    </div>
  )
}
