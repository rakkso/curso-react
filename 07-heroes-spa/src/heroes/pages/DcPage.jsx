import React from 'react'
import { HeroesList } from '../components/HeroesList'

export const DcPage = () => {
  return (
    <>
        <h1>DcPage</h1>
        <hr/>
    <HeroesList publisher='DC Comics'/>

    </>
  
  )
}
