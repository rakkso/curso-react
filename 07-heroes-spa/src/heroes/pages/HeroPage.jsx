import React from 'react'
import { HeroesList } from '../components/HeroesList'

export const HeroPage = () => {
  return (
    <>
    <h1>HeroPage</h1>
    
</>
    
  )
}
