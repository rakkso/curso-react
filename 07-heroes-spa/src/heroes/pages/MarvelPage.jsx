import React from 'react'
import { HeroesList } from '../components/HeroesList'

export const MarvelPage = () => {
  return (
    <>
    <h1>MarvelPage</h1>
    <hr/>
    <HeroesList publisher='Marvel Comics'/>
    </>
   
  )
}
