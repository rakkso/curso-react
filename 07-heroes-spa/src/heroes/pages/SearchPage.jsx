

import React from 'react'
import { HeroCard } from '../components/HeroCard'
import { useForm } from '../../hooks/useForm'
import { useLocation, useNavigate } from 'react-router-dom'

import { getHeroesByName } from '../helpers/getHeroesByName.js'
//import {queryString} from 'query-string'


export const SearchPage = () => {
  const queryString = require('query-string'); // con el import los test dan error
  const navigate = useNavigate();

  const location = useLocation();


  console.log("location %s",location);
  //const {query} =queryString.parse(location.search);
 // console.log(location);
  
  // Si descomento esto peta, no se el motivo : console.log("query:%s",query);
  const {miQuery=''} = queryString.parse(location.search);
  console.log("Mi query:%s",miQuery);
  const heroes = getHeroesByName(miQuery);
  console.log(heroes);
  const showSearch = (miQuery.length === 0);
  console.log (showSearch);
  const showError = (miQuery.length > 0) && heroes.length === 0;
  console.log (showError);
  

  const { searchText, onInputChange} =useForm({
    searchText:miQuery
  })

  const onSearchSubmit = (event) => {
    event.preventDefault();
    //console.log({searchText});

    //if (searchText.trim().length <=1 ) return;

    navigate(`?miQuery=${searchText}`);

  }

  return (
    <>
    <h1>SearchPage</h1> 
    <div className="row">
      <div className="col-5">
        <h4>Searching</h4>
        <hr/>
        <form onSubmit={onSearchSubmit}  aria-label="form">
          <input
            type="text"
            placeholder='Search a hero'
            className='form-control'
            name='searchText'
            autoComplete='off'
            value={searchText}
            onChange={onInputChange}
            />
            <button className='btn btn-outline-primary mt-1'>
              Search
            </button>
        </form>
      </div>
      <div className="col-7">
        <h4>Results</h4>
        <hr/>
        <div className="alert alert-primary animate__animated animate__fadeIn" style={{display: showSearch ? '':'none'}}>Search a hero</div>
        <div aria-label="alert-danger" className="alert alert-danger animate__animated animate__fadeIn" style={{display: showError ? '':'none'}}>No hero with <b>{miQuery}</b></div>

        {
          heroes.map(hero => (
            <HeroCard key={hero.id} {...hero} />
          ))
        }
        
      </div>
    </div>
    </>

   
  )
}
