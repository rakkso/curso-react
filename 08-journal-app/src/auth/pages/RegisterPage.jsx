import { Link as RouterLink } from 'react-router-dom';
import {useDispatch, useSelector} from 'react-redux'
import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material';
import { Google } from '@mui/icons-material';
import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks';
import { useMemo, useState } from 'react';
import { statCreatingUserWiehEmailPassword } from '../../store/auth/thunks';


const formData = {
  email: 'rak0kso@gmail.com',
  password : '123',
  displayName: 'Óscar' 
 }

 const formValidations = {
  email:[ (value) => value.includes('@'), 'El correo debe tener una @'],
  password: [(value)=> value.length >=6 ,'El password de de tener más de 6 letras'],
  displayName: [(value)=> value.length >=1,'El displayName es obligatorio']
}




export const RegisterPage = () => {

const dispatch = useDispatch();

const { estado, errorMessage} = useSelector(state => state.login);
const isCheckingAuthentication = useMemo( () => estado === 'checking' )


  const {
    formState, displayName, email, password, onInputChange,
    isFormValid, displayNameValid, emailValid, passwordValid
} = useForm(formData,formValidations);

console.log ("displayNameValid",displayNameValid);
const [formSubmitted,setFormSubmitted]= useState (false);
  const onSubmit = (event) =>{
    event.preventDefault();
    setFormSubmitted(true);
    console.log(formState);
    dispatch (statCreatingUserWiehEmailPassword(formState));
}

                {/* error={!displayNameValid}
                helperText = {displayNameValid}  */}

          
 
  return (
    <AuthLayout title="Crear cuenta">
      {
/*
<h1>FormValid: { isFormValid ? ' Válido':'Incorrecto' } </h1>*/

      }
      
      <form onSubmit={onSubmit}>
          <Grid container>
           
            <Grid item xs={ 12 } sx={{ mt: 2 }}>
              <TextField 
                label="Nombre completo" 
                type="text" 
                placeholder='Nombre completo' 
                fullWidth
                name="displayName"
                value={displayName}
                error = {!!displayNameValid && formSubmitted}
                helperText = {displayNameValid}

                onChange= {onInputChange}

              />
            </Grid>

            <Grid item xs={ 12 } sx={{ mt: 2 }}>
              <TextField 
                label="Correo" 
                type="email" 
                placeholder='correo@google.com' 
                fullWidth
                name="email"
                value={email}
                error = {!!emailValid && formSubmitted}
                helperText = {emailValid}

                onChange= {onInputChange}
              />
            </Grid>

            <Grid item xs={ 12 } sx={{ mt: 2 }}>
              <TextField 
                label="Contraseña" 
                type="password" 
                placeholder='Contraseña' 
                fullWidth
                name="password"
                value={password}
                error = {!!passwordValid && formSubmitted }
                helperText = {passwordValid}

                onChange= {onInputChange}
              />
            </Grid>
            <Grid container spacing={ 2 } sx={{ mb: 2, mt: 1 }}>

            <Grid item  xs={12}   display= { !!errorMessage ? '':'none'}>
                  <Alert severity='error' >{errorMessage}</Alert>
            </Grid>
            </Grid>
            
            <Grid container spacing={ 2 } sx={{ mb: 2, mt: 1 }}>
              <Grid item xs={ 12 }>
                <Button type="submit" variant='contained' fullWidth    display= { isCheckingAuthentication}>
                  Crear cuenta
                </Button>
              </Grid>
            </Grid>


            <Grid container direction='row' justifyContent='end'>
              <Typography sx={{ mr: 1 }}>¿Ya tienes cuenta?</Typography>
              <Link component={ RouterLink } color='inherit' to="/auth/login">
                ingresar
              </Link>
            </Grid>

          </Grid>


        </form>

    </AuthLayout>
  )
}
