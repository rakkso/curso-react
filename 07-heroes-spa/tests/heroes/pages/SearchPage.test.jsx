/**
 * @jest-environment jsdom
 */

import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchPage } from "../../../src/heroes/pages/SearchPage"


jest.mock('query-string', () => ({
    parse: jest.fn(() => ({ foo: 'bar', abc: 'xyz',miQuery:'batman' })),
    stringify: jest.fn(),
    // Add other methods you need to mock here
  }));

  const mockMifuncion = jest.fn();
//Mockea el useNavigate para poder preguntar si se ha ejecutado.
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate : () => mockMifuncion,
}))

describe('Pruebas de SearchPages', () => { 
    beforeEach(()=>jest.clearAllMocks()); //para dejar todos los mocks en el estado inicial.


    test('Deben mostarse los valores por defecto', () => { 
        
        const { container } = render(
            <MemoryRouter>
                <SearchPage/>
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();
     })

     test('Deben mostarse la busqueda de batman', () => { 
        
        const { container } = render(
            <MemoryRouter initialEntries={['/search?miQuery=batman']}>
                <SearchPage/>
            </MemoryRouter>
        )

        //screen.debug();
        const input = screen.getByRole('textbox');
        expect (input.value).toBe('batman');
        const img = screen.getByRole('img');
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg');
        const alert = screen.getByLabelText('alert-danger');
        expect (alert.style.display).toBe('none');


        
     })

     test('Debe mostar error si no encuenta', () => { 
        
        jest.clearAllMocks(); //para dejar todos los mocks en el estado inicial.
        jest.mock('query-string', () => ({
            parse: jest.fn(() => ({ foo: 'bar', abc: 'xyz',miQuery:'batman123' })),
            stringify: jest.fn(),
            // Add other methods you need to mock here
          }));


          render(
            <MemoryRouter initialEntries={['/search?miQuery=batman123']}>
                <SearchPage/>
            </MemoryRouter>

          )
          //screen.debug();

          //Este no lo hago porque como  he forzado el comportamiento del queryString a que retorne batman, no puedeo hacer que retorne otra cosa.
      })

      test('Cuando enviemos el formulario debe navegar', () => { 
        
        const inputValue='superman'
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage/>
            </MemoryRouter>

          )

          const input = screen.getByRole('textbox');
          fireEvent.change(input, {target: {name:'searchText',value:inputValue}});

          const form = screen.getByRole('form');
          fireEvent.submit(form);

          expect( mockMifuncion).toHaveBeenCalledWith(`?miQuery=${inputValue}`);



       })
 })
