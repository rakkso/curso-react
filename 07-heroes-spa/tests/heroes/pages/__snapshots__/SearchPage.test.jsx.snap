// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de SearchPages Deben mostarse los valores por defecto 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value="batman"
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
        style="display: none;"
      >
        Search a hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b>
          batman
        </b>
      </div>
      <div
        class="col animate__animated animate__fadeIn"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4 "
            >
              <img
                alt="Batman"
                class="card-img"
                src="/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-tittle"
                >
                  Batman
                </h5>
                <p
                  class="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  class="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                >
                  Mas ...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
